imports/wasi_snapshot_preview1/testdata/gotip/wasi.go

```diff
@@ -43,3 +50,64 @@ func mainSock() error {
 	fmt.Println(string(buf[:n]))
 	return nil
 }
+
+// mainHTTP implicitly tests non-blocking sockets, as they are needed for
+// middleware.
+func mainHTTP() error {
+	// Get the file representing a pre-opened TCP socket.
+	// The socket (listener) is the first pre-open, with a file-descriptor of
+	// 3 because the host didn't add any pre-opened files.
+	listenerFD := 3
+	f := os.NewFile(uintptr(listenerFD), "")
+
+	// Wasm runs similarly to GOMAXPROCS=1, so multiple goroutines cannot work
+	// in parallel. non-blocking allows the poller to park the go-routine
+	// accepting connections while work is done on one.
+	if err := syscall.SetNonblock(listenerFD, true); err != nil {
+		return err
+	}
+
+	// Convert the file representing the pre-opened socket to a listener, so
+	// that we can integrate it with HTTP middleware.
+	ln, err := net.FileListener(f)
+	defer f.Close()
+	if err != nil {
+		return err
+	}
+	defer ln.Close()
+
+	// Serve middleware that echos the request body to the response.
+	if err = http.Serve(&onlyOnceListener{Listener: ln}, echo{}); err == nil || err == http.ErrServerClosed {
+		return nil
+	} else {
+		return err
+	}
+}
+
+type onlyOnceListener struct {
+	net.Listener
+
+	accepted bool
+}
+
+func (ln *onlyOnceListener) Accept() (net.Conn, error) {
+	// TODO: this is a problem because we stop before the in-flight request is handled.
```

Comment:
the main issue in this test is whether we do it in the guest, or externally, we want the test to stop after one request. Currently, the TCP test stops after one request, and I wanted to do the same here.

Right now, this approach doesn't work because it closes the server before the in-flight request completes. I also tried adding a channel, but go into a deadlock and feel maybe I'm digging a hole.

Reply:
Actually I think the request is never handled. I'll revert the "only once" stuff, so that we can tell if anything is working first.